### PR TITLE
[Feature] Max readout and consecutive labeling for networkx

### PR DESCRIPTION
**Pull request by mufeili**
## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR

## Changes
- As requested in #306 , a naive implementation for max readout on nodes and edges is provided with `max_nodes` and `max_edges` in `batched_graph.py`. A more efficient implementation will require `segment_max`. The test cases are provided in `tests/pytorch/test_readout.py`
- As discussed in #334 , if we want to create a `DGLGraph` object from a networkx graph with non-consecutive node labels, we relabel the nodes using consecutive integers first before converting it into a `DGLGraph`. The changes are `from_networkx` in `graph.py`, `graph_index.py`, `immutable_graph_index.py`. The test cases are provided in `tests/compute/test_basics.py`.
- In `from_networkx` of `graph.py`, I also convert the networkx graph to be directed first, otherwise only half of the `DGLGraph` edges will have features.

### CODE REVIEW

**Comment by yzh119** [comment]
@mufeili does the `segment_max` you mentioned similar to `segment_sum` in tensorflow?

**Comment by mufeili** [comment]
@yzh119 Yeah, see [doc](https://www.tensorflow.org/api_docs/python/tf/math/segment_max) and [code](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/segment_reduction_ops.cc).

**Comment by BarclayII** [comment]
I think later on we will need to replace the for loop in `max_nodes` etc with a custom op, e.g. `scatter_max`.

**Comment by mufeili** [comment]
@BarclayII I agree, this is just a very naive workaround.

File: python/dgl/graph.py [hunk 1]
```diff
@@ -1165,6 +1170,10 @@ def from_networkx(self, nx_graph, node_attrs=None, edge_attrs=None):
                 [2., 2., 2., 2.],
                 [1., 1., 1., 1.]])
         """
+        nx_graph = nx_graph.to_directed()
+        # Relabel nodes using consecutive integers
+        nx_graph = nx.convert_node_labels_to_integers(nx_graph)
```
Comment (jermainewang) [inline python/dgl/graph.py]: Not needed. The graph index already calls it (see below).

File: python/dgl/graph.py [hunk 2]
```diff
@@ -1137,7 +1141,8 @@ def from_networkx(self, nx_graph, node_attrs=None, edge_attrs=None):
         Parameters
```
Comment (jermainewang) [inline python/dgl/graph.py]: Describe if the reordering happens, what ordering will be used. From nx: "inherit node ordering from G.nodes() “sorted” : inherit node ordering from sorted(G.nodes())"

File: python/dgl/graph.py [hunk 3]
```diff
@@ -1165,6 +1170,10 @@ def from_networkx(self, nx_graph, node_attrs=None, edge_attrs=None):
                 [2., 2., 2., 2.],
                 [1., 1., 1., 1.]])
         """
+        nx_graph = nx_graph.to_directed()
```
Comment (jermainewang) [inline python/dgl/graph.py]: What will happen for the node/edge attributes after `to_directed()`?

File: python/dgl/graph.py (hunk 1 above)
Comment (mufeili) [reply python/dgl/graph.py]: When we convert networkx graph into dgl graph with node features, it assumes that nodes are labeled with consecutive integers due to `for nid in range(self.number_of_nodes()):` in line 1193. Initially I thought relabeling the graph will help circumvent not only this issue for node feature but also for edge feature. It turns out that we cannot circumvent line 1208 anyway and I guess you are right that we should remove line 1173 and change line 1193 to be `for nid in nx_graph.nodes()` instead.

File: python/dgl/graph.py (hunk 3 above)
Comment (mufeili) [reply python/dgl/graph.py]: `to_directed` returns a deep copy of the original undirected graph where all `(node, node_attrs)` and `(edge, edge_attrs)` pairs are preserved. Besides, originally the undirected graph only has `u, v` and its associated attributes. `to_directed` creates `v, u` edges whose attributes are the same as `u, v`. See [`to_directed` source code](https://github.com/networkx/networkx/blob/master/networkx/classes/graph.py#L1602). This is useful as otherwise we will get `None` attributes for the new `v, u` edges and an error will occur in [L1178 of graph.py](https://github.com/dmlc/dgl/blob/master/python/dgl/graph.py#L1178) as you are concatenating tensors with None type objects.

Comment (mufeili) [reply python/dgl/graph.py]: I will add a condition check as the function seems still returns a deepcopy even if the graph is already directed.

Comment (jermainewang) [reply python/dgl/graph.py]: Got that. Maybe add some comments about this behavior?

File: python/dgl/graph.py (hunk 1 above)
Comment (jermainewang) [reply python/dgl/graph.py]: > It turns out that we cannot circumvent line 1208 anyway

Why is that? Also, technically `for nid in nx_graph.nodes()` does not assume any order so `for nid in range(self.number_of_nodes())` is clearer.

File: python/dgl/graph.py (hunk 3 above)
Comment (mufeili) [reply python/dgl/graph.py]: Comments added. I also added the condition of checking whether a networkx graph is directed in `graph_index.py` and `immutable_graph_index.py`.

File: python/dgl/graph.py (hunk 1 above)
Comment (mufeili) [reply python/dgl/graph.py]: 1. @ylfdq1118 Could you please take a look at whether my understanding is correct for the rest of item 1. From [L1196-1199 of graph.py](https://github.com/dmlc/dgl/blob/master/python/dgl/graph.py#L1196), we assume that there exists a prior specification of edge ids and the edge attributes are stored in the rows specified by these edge ids. @jermainewang Then if the edge ids are broken after we remove some nodes there will still be a problem. I guess this is not something that we need to take care.
2. If you do `for nid in range(self.number_of_nodes())`, since we did not relabel the nodes with consecutive integers in the `from_networkx` in `graph.py`, a KeyError can occur. I can do `for nid in sorted(nx_graph.nodes())` if this works for you.

Comment (jermainewang) [reply python/dgl/graph.py]: I see what happens here. Then, how about this?
(1) Relabel the networkx in `graph.py` because the order is required to set node features.
(2) `graph_index.py` will not relabel the nx graph and assume it is correct.

The benefit is we don't need to sort the nodes twice. As long as the `graph_index.from_networkx` is only called by `graph.from_networkx`, it would be fine. What do you think?

Comment (mufeili) [reply python/dgl/graph.py]: Sounds good to me. Done.

File: python/dgl/graph.py [hunk 4]
```diff
@@ -1181,7 +1194,7 @@ def _batcher(lst):
         if node_attrs is not None:
             # mapping from feature name to a list of tensors to be concatenated
             attr_dict = defaultdict(list)
-            for nid in range(self.number_of_nodes()):
+            for nid in nx_graph.nodes():
```
Comment (jermainewang) [inline python/dgl/graph.py]: Then revert this since the graph has been relabeled.

File: python/dgl/graph.py (hunk 1 above)
Comment (jermainewang) [reply python/dgl/graph.py]: For the edge id, maybe we could loop over the list and check whether they is a `None`? and if so, we raise error about the consecutive edge id blah blah.

File: python/dgl/graph.py (hunk 4 above)
Comment (mufeili) [reply python/dgl/graph.py]: Done.

File: python/dgl/graph.py (hunk 1 above)
Comment (mufeili) [reply python/dgl/graph.py]: Yeah I can do that. `attrs['id']` can also lead to an index out of range error, shall we also handle that?

Comment (jermainewang) [reply python/dgl/graph.py]: Yeah, that's a good idea.

Comment (mufeili) [reply python/dgl/graph.py]: Done.

**Comment by jermainewang** [comment]
Great! Thanks Murphy.
